F compression function implementation for Blake2b

pdyraga opened the pull request:
Closes: #1 

`F` is a compression function for Blake2. It takes as an argument the
state vector `h`, message block vector `mb`, offset counter `t`,
final block indicator flag `f`, and the number of rounds `rounds` to execute.
The state vector is modified in-place by the function. 

The code has been borrowed from `golang/crypto/blake2b` and adjusted so
that this function can be safely exposed as public. Parameters have
been also adjusted to match those in [RFC 7693](https://tools.ietf.org/html/rfc7693)

Test vectors were generated from those in `golang/crypto/blake2b`

pdyraga committed:
F compression function implementation for Blake2

F is a compression function for Blake2. It takes as an argument the
state vector h, message block vector blocks, 2-bit offset counter t,
final block indicator flag f, and the number of rounds to execute.
The state vector is modified in-place by the function. Number of rounds
can be anything from 0 to 12 (inclusive).

The code has been borrowed from golang/crypto/blake2b and adjusted so
that it can support any number of rounds between 0-12. Parameters have
been also adjusted to match those in RFC 7693

https://tools.ietf.org/html/rfc7693

Test vectors were generated from those in golang/crypto/blake2b.
The current test coverage is certainly not enough and we'll add some
additional test vectors in the future.

pdyraga commented on f.go:
```diff
@@ -0,0 +1,206 @@
+// block indicator flag `f`, and number of rounds `rounds`. The state vector
+// is modified by the function. Number of rounds can be anything between 0 and
+// 12 (inclusive).
+func F(h *[8]uint64, blocks []byte, t *[2]uint64, f uint64, rounds uint8) error {
```
Is [0,12] rounds enough for all our use cases?

pdyraga commented on f_test.go:
```diff
@@ -0,0 +1,63 @@
+	}
+}
+
+var testVectors = []struct {
```
We need some better test vectors, ideally something from Zcash.

pdyraga commented on f.go:
```diff
@@ -0,0 +1,206 @@
+
+const (
+	// BlockSize is the blocksize of BLAKE2 in bytes.
+	BlockSize = 128
```
`BlockSize` should be always `128` as far as I understand. Is that correct?

pdyraga committed:
Added a bunch of F compression function test vectors

Those test vectors were generated from golang.org/x/crypto/blake2b test
hashes in the testHashashes function before and after each F execution.

pdyraga committed:
Added information about KEEP SEZC modifications

mhluongo commented on f.go:
```diff
@@ -0,0 +1,208 @@
+		}
+
+		for j := 0; j < int(rounds); j++ {
+			s := &(precomputed[j])
```
We can support up to `2 ^ 8 - 1` rounds if we replace this `j` with `j mod 10` and get rid of the last two values in the `precompute` array. It'll be at a slight performance expense but 🤷‍♂️

pdyraga replied:
Done!

mhluongo reviewed (commented):
Interested in your thoughts on raising the round limit to make this more general @pdyraga

mhluongo commented:
Looks like this will close #1 and #3 as well

pdyraga committed:
Allow for an arbitrary number of F compression rounds

pdyraga committed:
Accept last block flag f as boolean

Just like in the original RFC
(https://tools.ietf.org/html/rfc7693#section-3.2), f parameter is now
represented as boolean. If it is enabled, bits are inverted.

pdyraga committed:
Additional test vectors generated from TestHashes2X

We use test vectors generated from test functions in
golang.org/x/crypto/blake2b. Each unique call to hashBlocks has been
transformed into a test vector.

Here, we add test vectors from TestHashes2X test function.
We also moved test vectors generated from TestHashes test function into
a separate file so that test vectors generated from individual test
functions are stored separately.

pdyraga committed:
Fixed iterator bug - we should iterate over blocks, not by rounds

I mistakenly changed it in one of the previous commits - fixing it now.

pdyraga committed:
Introduced testVector struct and simplified test cases

Extracted common test code to a separate function.

pdyraga committed:
Ensure message block vector length is as expected

Message block vector length has to match block size (128) so that we
don't blow up when evaluating m vector later. In the original
implementation, F was an internal function and other public functions
calling it were making sure the correct vector is passed. Here, after
making F public, we need to carve this requirement in the interface.

pdyraga committed:
Updated repository link in go.mod

pdyraga commented:
@mhluongo All comments addressed, I am happy with what's here. It's ready for your review 👀.

pdyraga commented:
#3 will be covered in go-ethereum PR - each contract has benchmarks there.

mhluongo commented on f.go:
```diff
@@ -0,0 +1,199 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Modified by KEEP SEZC to expose F compression function.
```
Going to touch this up after this PR

pdyraga replied:
Yeah, I had no idea what to put here. Worth double-checking if we do not have to alter `LICENSE` as well.

mhluongo reviewed (approved):
:rocket: :rocket: :rocket:

pdyraga commented:
:tada:

carterpy commented:
hey @pdyraga here are those questions: 

@pdyraga, I've been playing with your f function in go and it's great and you've done a fantastic job here - massive props. I have two questions but please do not answer the stupid ones:

    In F.go, should f return h?
    Should the package be the same name as the folder it's in?
